pkg/ruler/ruler.go

```diff
@@ -347,21 +349,32 @@ func (r *Ruler) run(ctx context.Context) error {
 	tick := time.NewTicker(r.cfg.PollInterval)
 	defer tick.Stop()
 
-	r.loadRules(ctx)
+	r.loadRules(ctx, "")
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
 		case <-tick.C:
-			r.loadRules(ctx)
+			r.loadRules(ctx, "")
 		}
 	}
 }
 
-func (r *Ruler) loadRules(ctx context.Context) {
+// Load the rules for a particular userID. If userID empty, we load all rules.
+func (r *Ruler) loadRules(ctx context.Context, userID string) {
 	ringHasher := fnv.New32a()
 
-	configs, err := r.store.ListAllRuleGroups(ctx)
+	configs := map[string]rules.RuleGroupList{}
+	var err error
+
+	if userID == "" {
+		configs, err = r.store.ListAllRuleGroups(ctx)
+	} else {
+		var rgs rules.RuleGroupList
+		rgs, err = r.store.ListRuleGroups(ctx, userID, "")
+		configs[userID] = rgs
+	}
+
```

Comment:
This is dangerous. `r.manager.SyncRuleGroups` always needs *All* users to be present in the map, with all of the rules. If it doesn't find the user in the supplied map, rules manager for given user is stopped and removed from memory.